src/lib.ml

```diff
@@ -864,48 +887,39 @@ let rec instantiate db name info ~stack ~hidden =
     resolve_dep db (name : Lib_name.t) ~allow_private_deps ~loc ~stack in
 
   let implements =
-    Lib_info.implements info
-    |> Option.map ~f:(fun ((loc,  _) as name) ->
-      let* vlib = resolve name in
-      let virtual_ = Lib_info.virtual_ vlib.info in
-      match virtual_ with
-      | None ->
-        Error.not_virtual_lib ~loc ~impl:info ~not_vlib:vlib.info
-      | Some _ ->
-        let variant = Lib_info.variant info in
-        match variant with
-        | None -> Ok vlib
-        | Some variant ->
-          (* If the library is an implementation tagged with a
-             variant, we must make sure that that it implements a
-             library that is part of the same project *)
-          let status_vlib = Lib_info.status vlib.info in
-          if Option.equal Dune_project.Name.equal
-               (Lib_info.Status.project_name status)
-               (Lib_info.Status.project_name status_vlib)
-          then
-            Ok vlib
-          else
-            let name = Lib_info.name info in
-            let name_vlib = Lib_info.name vlib.info in
-            User_error.raise ~loc
-              [ Pp.textf
-                  "Library implementation %s with variant %s implements a \
-                   library outside the project. Instead of using \
-                   (variant %S) here, you need to reference it in the \
-                   virtual library project, using the external_variant stanza:"
-                  (Lib_name.to_string name)
-                  (Variant.to_string variant)
-                  (Variant.to_string variant)
-              ; Pp.textf
-                  "(external_variant\n\
-                  \  (virtual_library %s)\n\
-                  \  (variant %S)\n\
-                  \  (implementation %s))"
-                  (Lib_name.to_string name_vlib)
-                  (Variant.to_string variant)
-                  (Lib_name.to_string name)
-              ])
+    let open Option.O in
+    let+ ((loc, _) as name) = Lib_info.implements info in
+    let open Result.O in
+    let* vlib = resolve name in
+    let virtual_ = Lib_info.virtual_ vlib.info in
+    match virtual_ with
+    | None ->
+      Error.not_virtual_lib ~loc ~impl:info ~not_vlib:vlib.info
+    | Some _ ->
+      let variant = Lib_info.variant info in
+      begin match variant with
+      | None -> Ok vlib
+      | Some variant ->
+        (* If the library is an implementation tagged with a variant, we must
+           make sure that that it's correctly part of the virtual library's
+           known implementations. *)
+        let name = Lib_info.name info in
+        let error () =
+          let vlib_name = Lib_info.name vlib.info in
+          Error.vlib_known_implementation_mismatch
+            ~loc ~name ~variant ~vlib_name
+        in
+        let vlib_impls = Lib_info.known_implementations vlib.info in
+        let* (_, impl_name) =
+          match Variant.Map.find vlib_impls variant with
+          | None -> error ()
```

Comment:
I've refactored the code a bit, and now it seems that the same error is used for both cases. Do you think we should perhaps use more descriptive errors for both cases?

Reply:
Indeed, we can separate the errors to make them more specific:
* one case is when the variant is used for another library
* the other is when the library doesn't know about the variant